mm/map/mm_map.c

```diff
@@ -0,0 +1,327 @@
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/mm/map.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <nuttx/sched.h>
+#include <nuttx/kmalloc.h>
+#include <assert.h>
+
+#if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static bool in_range(FAR const void *start, size_t length,
+                     FAR const void *range_start, size_t range_length)
+{
+  char *u_start = (char *)start;
+  char *u_end = u_start + length;
+  char *r_start = (char *)range_start;
+  char *r_end = r_start + range_length;
+
+  return (u_start >= r_start && u_start < r_end && /* start is in range */
+          u_end >= r_start && u_end <= r_end);     /* end is in range */
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: mm_map_lock
+ *
+ * Description:
+ *   Get exclusive access to task_group's mm_map
+ *
+ ****************************************************************************/
+
+int mm_map_lock(void)
+{
+  return nxrmutex_lock(&nxsched_self()->group->tg_mm_map.mm_map_mutex);
+}
+
+/****************************************************************************
+ * Name: mm_map_unlock
+ *
+ * Description:
+ *   Relinquish exclusive access to task_group's mm_map
+ *
+ ****************************************************************************/
+
+void mm_map_unlock(void)
+{
+  DEBUGVERIFY(
+    nxrmutex_unlock(&nxsched_self()->group->tg_mm_map.mm_map_mutex)
+  );
+}
+
+/****************************************************************************
+ * Name: mm_map_initialize
+ *
+ * Description:
+ *   Allocates a task group specific mm_map stucture. Called when the group
+ *   is initialized
+ *
+ ****************************************************************************/
+
+void mm_map_initialize(FAR struct mm_map_s *mm)
```

Comment:
why not get from the current task

Reply:
Correct me if I am wrong

This is called during group initialization. The task is not yet running, but current task returns g_readytorun.head ( or in SMP case the g_assignedtasks[cpu].head)

The task cannot be current/running yet if the group is only being initialized... But you could pass group here if it looks better.

Then you have group in two functions, which are required when the group is being created and destroyed.

Reply:
so, that's why I prefer the first approach, which is more consistent. There are many exceptions you can't use the current task.

Reply:
Initializing and destroying the map is correct I think. Only the mm_map_s structure is passed, which is the one to initialize (in task group creation) and destroyed (in task group destruction, process exit)

The only question is that how you pass the info about task exit case to driver's unmap (from the mm_map_destroy). You can either have the logic in every driver, conditionally calling mm_map_remove only in normal case(not during group destruction) OR forwarding this info to mm_map_remove.

Reply:
> Initializing and destroying the map is correct I think. Only the mm_map_s structure is passed, which is the one to initialize (in task group creation) and destroyed (in task group destruction, process exit)
> 

Here is three choices:

1. Get info from the current task
2. Pass task_group_s as first argument
3. Pass mm_map_s as first argument

Please don't mix the usage as much as you can, any exception need have a very good reason.

> The only question is that how you pass the info about task exit case to driver's unmap (from the mm_map_destroy). You can either have the logic in every driver, conditionally calling mm_map_remove only in normal case(not during group destruction) OR forwarding this info to mm_map_remove.
> 
> If passing group for this purpose feels wrong, we can
> 
> * pass mm_map_s instead (NULL or current one)
> * a boolean flag
> * we can conditionally call the mm_map_remove from drivers only when the context is known (not during process exit).
> 
> We can even re-think the driver's munmap "group" argument. That was given only because it is a good indication to the driver that the context is not valid when the group is null.

Reply:
Getting info from current task. Accessing group struct belongs in sched, so added a macros into sched.h to get the current mm and mm from group.